feat(trade): sync trade amounts between widgets

shoom3301 opened the pull request:
# Summary

<img width="511" alt="image" src="https://github.com/cowprotocol/cowswap/assets/7122625/cb6b8f86-9a86-4033-8c27-d23278eeb053">

Problem:
When you open swap and set: Sell 120 COW for USDC
Then navigate to Limit orders
You expect to sell Sell 120 COW in limit orders widget as well, but you don't.

In this PR I've added persistance of trade amounts and kind into URL.
Having this, we can restore the state after navigation to another widget.

There is one problem with TWAP, because it doesn't support buy orders, so we ignore `orderKind` parameter from URL and always set only sellAmount.

# To Test

1. Set some trade with amounts
2. Navigate to another trade widget using the widget trade menu (see the pic above)
- [ ] currencies and amounts should be the same as before navigation
3. Repeat with different directions, amounts, order types

alfetopito commented on apps/cowswap-frontend/src/modules/trade/const/tradeUrl.ts:
```diff
@@ -1,2 +1,3 @@
 export const TRADE_URL_SELL_AMOUNT_KEY = 'sellAmount'
 export const TRADE_URL_BUY_AMOUNT_KEY = 'buyAmount'
+export const TRADE_URL_ORDER_KIND_KEY = 'orderKind'
```
Can we make it shorter?

```suggestion
export const TRADE_URL_ORDER_KIND_KEY = 'kind'
```

shoom3301 replied:
But why? `kind` sounds too generic

alfetopito replied:
Just because I like it shorter.
OrderKind feels too verbose.
But I'm fine if you want to keep it.

alfetopito reviewed (commented):
2 issues that I found

Decimals:

https://github.com/cowprotocol/cowswap/assets/43217/3890ee71-0608-4a66-858f-e3f09a18ae37

Buy orders:

https://github.com/cowprotocol/cowswap/assets/43217/81cebd5e-83ff-410f-93c2-cd26ce03927c


https://github.com/cowprotocol/cowswap/assets/43217/cab47277-2f2e-47ac-ba99-d639c4dcca90

elena-zh commented:
Hey @shoom3301 , besides the issues @alfetopito reported above, there are some more: 
1. Go to TWAP page
2. Select a token pair and an amount to sell
3. Then navigate to the Limit orders page --> sell amount is not valid 
![image](https://github.com/cowprotocol/cowswap/assets/70885163/c630b27e-5351-4168-9857-778b378239cd)
![image](https://github.com/cowprotocol/cowswap/assets/70885163/037f6bf8-6a91-4c63-83cd-dcf972dbb40c)
-----
order type is not correct in the link when: 
1. Open swap page
2. Switch token pair: sell token empty, buy token = weth, weth amount is filled in
3. navigate to the TWAP page
4. navigate to Limit orders
![from twap to limit](https://github.com/cowprotocol/cowswap/assets/70885163/1c5f2f29-07d8-47a7-bca1-3f7d3a584680)
---
 amounts in the link are not transferred/wrongly transfeerred to the limit orders page from Swap (buy orders): 
![image](https://github.com/cowprotocol/cowswap/assets/70885163/5de021cf-5009-4fec-b1d0-6eeb388f7085)
![buy order](https://github.com/cowprotocol/cowswap/assets/70885163/a5fa8999-136b-470b-8950-db3bc904df28)

anxolin reviewed (commented):
Great initiative!

Other than the reported issues, when I changed the tab the buy amount was not calculated automatically even though there was a price and there's a sell amount.

<img width="1320" alt="Screenshot at Jun 11 09-57-39" src="https://github.com/cowprotocol/cowswap/assets/2352112/add68ed4-ad1f-4d47-a0c4-baa9ae0a1b41">

shoom3301 commented:
Thank you guys! Fixed all issues.

@elena-zh about this

>order type is not correct in the link when:

TWAP doesn't support buy orders, so the flow might be a bit unclear. Anyway, I implemented it as:
1. Open /limit/_/WETH and fill 55 as buy amount
2. Go to Swap and fill 33 as buy amount
3. Go to Twap. Both sell and buy amount should be zero
4. Go to Limit. You should see 55, because you came from TWAP where we didn't have amounts (both are zero) so we take previous remembered state

alfetopito reviewed (approved):
Working great!

elena-zh commented:
> Thank you guys! Fixed all issues.
> 
> @elena-zh about this
> 
> > order type is not correct in the link when:
> 
> TWAP doesn't support buy orders, so the flow might be a bit unclear. Anyway, I implemented it as:
> 
> 1. Open /limit/_/WETH and fill 55 as buy amount
> 2. Go to Swap and fill 33 as buy amount
> 3. Go to Twap. Both sell and buy amount should be zero
> 4. Go to Limit. You should see 55, because you came from TWAP where we didn't have amounts (both are zero) so we take previous remembered state

Hey @shoom3301 , great job overall!
One more edge case: 
![image](https://github.com/cowprotocol/cowswap/assets/70885163/df2451dd-2514-4ea0-b2a6-70b175f46bed)

1. Open /limit/_/WETH and fill 55 as buy amount
![image](https://github.com/cowprotocol/cowswap/assets/70885163/aaa992e4-a354-4107-825b-88cda4609c7c)

2. Go to Swap and fill in empty token sell field with 33
![image](https://github.com/cowprotocol/cowswap/assets/70885163/408e6958-b7d3-4c4e-934d-285515e07dd2)

3. Go to TWAP --> 0
![image](https://github.com/cowprotocol/cowswap/assets/70885163/c755f61b-0d25-486e-a597-997fac494ea8)

4. Go to Swap --> zeros instead of 33 in the sell field is displayed
![image](https://github.com/cowprotocol/cowswap/assets/70885163/7b9ab99a-4481-45d2-adba-7f8980b58b27)

5. Pick ETH in the sell field --> 33 will appear
![image](https://github.com/cowprotocol/cowswap/assets/70885163/cb25dd8f-9ba4-41bf-b75e-7324d35ef188)

6. Go to TWAP --> IDK why, but I see 500 in the TWAP wrap
![image](https://github.com/cowprotocol/cowswap/assets/70885163/aab2c730-620a-45c4-82f1-3b2b51b15058)

So something is weird happening here. Anyways, LMK if I need to open a separate issue for this. 

Thanks

shoom3301 commented:
@elena-zh let's skip this super edge-case, I don't think we should spend time on it.
But anyway, thanks for attention :)
